kpathsea/src/lib.rs

```diff
@@ -19,13 +19,72 @@ impl Kpaths {
     Kpaths(kpse)
   }
 
+  /// For a given filename, try to guess the kpse format type from the file
+  /// extension by looking it up in the format info table. This is a simplified
+  /// version of the find_format function in kpsewhich.
+  fn guess_format_from_filename(&self, filename: &str) -> kpse_file_format_type {
+    // We go through each format type
+    for format_type in 0..kpse_file_format_type_kpse_last_format {
+      let format_info = unsafe { (*self.0).format_info[format_type as usize] };
+
+      if format_info.type_.is_null() {
+        // If this format hasn't been initialized yet, initialize it now.
+        // Otherwise, it won't have the list of suffixes initialized.
+        unsafe {
+          kpathsea_init_format(self.0, format_type as kpse_file_format_type);
+        }
+      }
+
+      // First, we check the suffixes for each format type. The suffixes are
+      // stored as an array of strings with a null pointer denoting the last
+      // value. Also, the pointer to the array can itself be null if there are
+      // no suffixes.
+      let mut suffix_ptr = format_info.suffix;
+      while !suffix_ptr.is_null() && !unsafe {*suffix_ptr}.is_null() {
+        // Pull out the suffix
+        let suffix_cstr = unsafe { CStr::from_ptr(*suffix_ptr) };
+        let suffix = suffix_cstr.to_str().unwrap();
```

Comment:
Looking at the suffixes reported locally, I am unsure about getting side-effects with some of the more unusual ones. So two fast ideas to upgrade before merging may be:
 - lowercasing both the kpathsea suffix and the result of `filename.get()` to ensure we're not getting a case mismatch issue
 - checking if the `suffix` starts with a `.` and ignoring it if not. Or figuring out what the `gf` and `pk` values are, which is the reason I am bringing this bit up. 

Here's the (sorted) list from the test run from a simple `dgb!()` added here:
```rust
let suffix = dbg!(suffix_cstr.to_str().unwrap());
```

```
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".afm"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".base"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".bib"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".bltxml"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".bst"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".cid"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".cidmap"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".cnf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".dfont"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".dll"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".enc"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".fea"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".fmt"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ist"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".lig"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".lua"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".luatex"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".luc"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".luctex"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".map"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".mem"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".mf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".mft"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".mlbib"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".mlbst"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".mp"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ocp"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ofm"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".opl"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".OTF"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".otf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".otp"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ovf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ovp"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".pfa"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".pfb"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".pool"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ris"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".sfd"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".so"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".t42"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".T42"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".tex"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".texlua"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".texluc"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".tfm"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".tlu"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".TTC"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ttc"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".ttf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".TTF"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".vf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".w"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = ".web"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = "gf"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = "ls-r"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = "ls-R"
[kpathsea/src/lib.rs:46] suffix_cstr.to_str().unwrap() = "pk"
```